### PR TITLE
feat(dpmodel): backend-agnostic DPA4 descriptor and fitting (core path)

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
## What this PR does

Ports the core DPA4/SeZM descriptor and its GLU energy fitting from the PyTorch-only implementation (`deepmd/pt`, #5448/#5503) to the backend-agnostic dpmodel layer (numpy / array-API). This is PR 1 of a 3-PR series (next: pt_expt wrappers + training; then `.pt2` freeze + Python/C++ inference).

### Contents

- `deepmd/dpmodel/utils/lebedev.py` + relocated `lebedev_rules.npz` (single source of truth; pt re-wraps it), `deepmd/dpmodel/utils/spherical_harmonics.py` — general-lmax real spherical harmonics in pure numpy, matching e3nn's `normalize=True, normalization="norm"` convention exactly (validated vs e3nn to ~1e-15), so dpmodel builds the S2-grid constants without e3nn/torch.
- `deepmd/dpmodel/descriptor/dpa4_nn/` — array-API ports of the `sezm_nn` math modules (indexing, utils, radial, norm, wignerd, so3, activation, projection, grid_net, so2, attention, embedding, edge_cache, ffn, block), structurally mirroring the pt code 1:1.
- `deepmd/dpmodel/descriptor/dpa4.py` — `DescrptDPA4`, registered `dpa4`/`DPA4`/`sezm`/`SeZM`, serialization-compatible with pt `DescrptSeZM` (cross-deserialization tested in both directions).
- `deepmd/dpmodel/fitting/dpa4_ener.py` — `SeZMEnergyFittingNet` (GLU fitting), registered `dpa4_ener`/`sezm_ener`.

### Design: padded edges instead of sparse `nonzero` edges

pt builds a sparse edge list via `torch.nonzero` and aggregates with `index_add_`/`scatter`. The dpmodel port uses a shape-static padded edge layout (`E = nf*nloc*nnei`, node-contiguous, validity mask): per-edge math is identical, node aggregations and the segment softmax become masked reductions over the `nnei` axis. This removes data-dependent shapes (torch.export-friendly for the upcoming pt_expt PR) and avoids non-standard scatter ops in array-API code. Masked slots carry a safe dummy direction before quaternion/Wigner evaluation so future autograd backward passes cannot be NaN-poisoned by dead slots.

### Validation

- Per-module weight-copy parity tests vs the pt classes (fp64, rtol 1e-12 / atol 1e-14) in `source/tests/pt/model/test_dpa4_dpmodel_parity.py` (377 tests), including dual-cache tests proving pt-sparse vs dp-padded equivalence with adversarial garbage in masked slots.
- Descriptor-level end-to-end parity vs pt: max abs error ~6e-15; descriptor+fitting chained parity at rtol 1e-10.
- Torch-free tests (`source/tests/common/dpmodel/`): serialize round-trips, permutation equivariance, rotation invariance, masked-edge inertness, guard coverage.
- Cross-backend harness tests: `source/tests/consistent/descriptor/test_dpa4.py`, `source/tests/consistent/fitting/test_dpa4_ener.py` (dpa3-precedent tolerances; f64 atol stricter than precedent).
- A model-dict contract test pins the pt `dpa4` model `serialize()` top-level shape for the follow-up PRs.

## Known limitations

- **Core path only.** Out-of-scope features raise `NotImplementedError` naming the config key at construction/deserialize: spin/charge embedding (`add_chg_spin_ebd`), DeNS dual fitting, ZBL/inner-clamp bridging, multitask FiLM/case embedding (`dim_case_embd>0`), LoRA, SO3-grid paths (`ffn_so3_grid`, `node_wise_*`, `message_node_*`), `grid_mlp`, attention-residual modes, `layer_scale`, attention projections.
- **Lebedev quadrature only**: `lebedev_quadrature=False` (e3nn product grid) is rejected; the default config uses Lebedev.
- **Wigner-D evaluates the closed form** that pt's lstsq-fitted monomial kernels approximate; parity ≤1e-14 for the core lmax≤4 range, diverging O(1e-13) only at lmax≥9 (pt's own kernels drift from the closed form there).
- **Padded edges spend compute on masked slots** (like dpa1/2/3); pt's sparse path remains faster for very sparse neighbor lists. `random_gamma` gauge augmentation is accepted and serialized but applied only when explicitly driven (numpy RNG; not stream-identical to torch's draw — gauge invariance is property-tested).
- **No model-level dpmodel assembly yet** (energy model wiring, pt_expt wrappers, training, freeze, inference are PR-2/PR-3).
- dpmodel numpy execution is a correctness reference, not a performance target; fp16/bf16 compute-dtype promotion points are not bit-mirrored (fp64/fp32 are).

## Summary by CodeRabbit

* **New Features**
  * Full DPA4/SeZM descriptor and building blocks added to the dpmodel backend (equivariant layers, radial/grid/Wigner support, and serialization/deserialize support).
  * DPA4 (SeZM) GLU-based energy fitting networks and network collection.
  * Packaged Lebedev quadrature loader and real spherical harmonics utilities.

* **Bug Fixes**
  * Device/array alignment fixes to prevent backend/device mismatches.

* **Tests**
  * Extensive torch-free and cross-backend tests for descriptor, fitting, Lebedev, and spherical-harmonics.

### CODE REVIEW

**Review by OutisLi — APPROVED** [review]

File: deepmd/dpmodel/descriptor/dpa4_nn/so2.py [hunk 1]
```diff
@@ -0,0 +1,1736 @@
...
+        out = xp.reshape(
+            out_flat, (n_edge, self.n_focus, self.reduced_dim, self.out_channels)
+        )
+
+        # === Step 3. Bias on l=0 scalar index ===
+        if self.mlp_bias:
+            bias0 = xp.reshape(
+                xp.asarray(self.bias0[...], device=device),
+                (self.n_focus, self.out_channels),
+            )
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/descriptor/dpa4_nn/so2.py:397]: All three sites fixed in 70702ce14: `so2_linear.bias0` in the bias-correction closure, `AtomExcludeMask.type_mask`, and `NativeLayer.idt` now go through `xp.asarray(..., device=...)` before runtime arithmetic.

File: deepmd/dpmodel/descriptor/dpa4_nn/wignerd.py [hunk 2]
```diff
@@ -0,0 +1,963 @@
...
+    -------
+    Array
+        Unit quaternions with shape ``(E, 4)`` in ``(w, x, y, z)`` order.
+    """
+    xp = array_api_compat.array_namespace(edge_vec)
+    if edge_len is None:
+        edge_len = safe_norm(edge_vec, eps)
+    else:
+        edge_len = xp.sqrt(edge_len * edge_len + eps * eps)
+    edge_unit = edge_vec / edge_len
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/descriptor/dpa4_nn/wignerd.py:249]: Rejecting: the `sqrt(edge_len**2 + eps**2)` re-flooring is exactly what the pt reference does in the same branch (deepmd/pt/model/descriptor/sezm_nn/wignerd.py:371). The dpmodel port must reproduce pt bit-for-bit (the parity tests pin this at rtol 1e-12); changing it here would break cross-backend consistency with trained pt checkpoints.

File: deepmd/dpmodel/descriptor/dpa4_nn/norm.py [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+
+        # Pre-fuse degree balancing and channel averaging into a single weight:
+        #   w_d = 1 / ((2l+1) * (lmax+1) * C)
+        # so that the shared RMS statistic is a single weighted sum without
+        # allocating an intermediate (N, D, F, C) buffer beyond x^2 itself.
+        weights_list = []
+        scale = 1.0 / ((self.lmax + 1) * self.channels)
+        for l in range(self.lmax + 1):
+            w = scale / (2 * l + 1)
+            weights_list.extend([w] * (2 * l + 1))
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/descriptor/dpa4_nn/norm.py:217]: Rejecting: E741 is globally ignored in this repo (pyproject.toml ruff ignore list), so `ruff check .` and pre-commit pass as-is. The single-letter `l` (degree index) also matches the pt sezm_nn code this file mirrors, which helps side-by-side review.

File: deepmd/dpmodel/descriptor/dpa4.py [hunk 4]
```diff
@@ -0,0 +1,1258 @@
...
+        # === L/M schedules ===
+        self._init_lm_schedules(lmax, n_blocks, l_schedule, mmax, m_schedule)
+        self.kmax = int(kmax)
+        if self.kmax < 0:
+            raise ValueError("`kmax` must be non-negative")
+        if self.kmax > self.lmax:
+            raise ValueError("`kmax` must be <= `lmax`")
+        self.ebed_dims = [get_so3_dim_of_lmax(l) for l in self.l_schedule]
+        self._init_node_l_schedules(extra_node_l)
+        self.rad_sizes_per_block = [l + 1 for l in self.l_schedule]
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/descriptor/dpa4.py:376]: Rejecting: same as the norm.py thread — E741 is in the repo's ruff ignore list, and `l` matches the pt sezm.py source for cross-referencing. CI lint passes.

File: deepmd/dpmodel/descriptor/dpa4_nn/embedding.py [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+        -------
+        Array
+            Type embeddings with shape (..., embed_dim).
+        """
+        xp = array_api_compat.array_namespace(atype)
+        weight = self.adam_type_embedding[...]
+        # pt embedding.py:143 torch.embedding -> flat int64 take + reshape.
+        index = xp.astype(xp.reshape(atype, (-1,)), xp.int64)
+        out = xp.take(weight, index, axis=0)
+        return xp.reshape(out, (*atype.shape, self.embed_dim))
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/descriptor/dpa4_nn/embedding.py:166]: Rejecting the runtime check: a data-dependent raise on tensor values would break torch.export tracing when pt_expt wraps this class (PR-2 of this series), and sibling dpmodel embeddings do not validate atype either. Negative ids are documented as a caller-contract violation in the call() docstring (added in 93e4d0b2d), matching pt where torch.embedding errors natively.

File: deepmd/dpmodel/fitting/dpa4_ener.py [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+                    f"key {key} contains type indices outside [0, {self.ntypes})"
+                )
+            idx = sum([tt * self.ntypes**ii for ii, tt in enumerate(key)])
+        if not (0 <= idx < self.ntypes**self.ndim):
+            raise ValueError(
+                f"key {key} maps to index {idx}, outside [0, {self.ntypes**self.ndim})"
+            )
+        return idx
+
+    def __getitem__(self, key: int | tuple | str) -> GLUFittingNet:
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/fitting/dpa4_ener.py:265]: Fixed in 70702ce14: `_convert_key`/`__getitem__` now raise KeyError for invalid/unset keys per the lookup-method convention.

**Comment by wanghan-iapcm** [comment]
pre-commit.ci run
